File: app/src/main/java/one/mixin/android/vo/ArrayConverters.java
package one.mixin.android.vo;

import androidx.room.TypeConverter;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;

import java.lang.reflect.Type;
import java.util.ArrayList;
import one.mixin.android.util.GsonHelper;

public class ArrayConverters {
    @TypeConverter
    public static ArrayList<String> fromString(String value) {
        Type listType = new TypeToken<ArrayList<String>>() {
        }.getType();
        return GsonHelper.INSTANCE.getCustomGson().fromJson(value, listType);
    }

    @TypeConverter
    public static String fromArrayList(ArrayList<String> list) {
        Gson gson = GsonHelper.INSTANCE.getCustomGson();
        return gson.toJson(list);
    }
}

File: app/src/main/java/one/mixin/android/vo/ArrayConverters.kt
package one.mixin.android.vo

import androidx.room.TypeConverter
import com.squareup.moshi.Types
import one.mixin.android.moshi.MoshiHelper.getTypeAdapter
import java.util.ArrayList

object ArrayConverters {
    @JvmStatic
    @TypeConverter
    fun fromString(value: String?): ArrayList<String> {
        if (value == null) return arrayListOf()
        return requireNotNull(jsonAdapter.fromJson(value))
    }

    @JvmStatic
    @TypeConverter
    fun fromArrayList(list: ArrayList<String>?): String {
        return jsonAdapter.toJson(list)
    }

    private val jsonAdapter by lazy {
        val listType = Types.newParameterizedType(ArrayList::class.java, String::class.java)
        getTypeAdapter<ArrayList<String>>(listType)
    }
}
